core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -2169,57 +2142,106 @@ public static boolean equalType(String desc0, MutableRel rel0, String desc1,
-      @Nullable List<RexNode> leftProjects, @Nullable List<RexNode> rightProjects) {
-    if (leftProjects != null && joinType.generatesNullsOn(0)
-        && !allProjectsNullable(leftProjects)) {
-      return false;
-    }
-    if (rightProjects != null && joinType.generatesNullsOn(1)
-        && !allProjectsNullable(rightProjects)) {
-      return false;
+  private static List<RexNode> shiftAndAdjustProjectExpr(MutableJoin query, MutableJoin target,
+      RexBuilder rexBuilder, int leftCount,
+      @Nullable List<RexNode> qInput0Projs, @Nullable List<RelDataTypeField> qInput0InputFields,
+      @Nullable List<RexNode> qInput1Projs, @Nullable List<RelDataTypeField> qInput1InputFields) {
+
+    int[] adjustments0 = new int[target.rowType.getFieldCount()];
+    int[] adjustments1 = new int[target.rowType.getFieldCount()];
+
+    if (qInput1Projs != null) {
+      Arrays.fill(adjustments1, fieldCnt(target.getLeft()));
+    }
+
+    // In cases such as JoinOnLeftCalcToJoinUnifyRule and JoinOnCalcsToJoinUnifyRule rules,
+    // the left calc need to be pulled up the target, initialize the converter.
+    RelOptUtil.RexInputConverter converter0 =
+        new RelOptUtil.RexInputConverter(rexBuilder, qInput0InputFields,
+            target.rowType.getFieldList(), adjustments0);
+
+    // In cases such as JoinOnRightCalcToJoinUnifyRule and JoinOnCalcsToJoinUnifyRule rules,
+    // the right calc need to be pulled up the target, initialize the converter.
+    RelOptUtil.RexInputConverter converter1 =
+        new RelOptUtil.RexInputConverter(rexBuilder, qInput1InputFields,
+            target.rowType.getFieldList(), adjustments1);
+
+    final List<RexNode> compenProjs = new ArrayList<>();
+    for (int i = 0; i < fieldCnt(query); i++) {
+      RelDataType type = query.rowType.getFieldList().get(i).getType();
+      if (i < leftCount) {
+        if (qInput0Projs == null) {
+          compenProjs.add(new RexInputRef(i, type));
+        } else {
+          RexNode apply = converter0.apply(qInput0Projs.get(i));
+          compenProjs.add(adjustNullability(apply, type, rexBuilder));
+        }
+      } else {
+        if (qInput1Projs == null) {
+          // This is equivalent to a shift, but without a calc on the right side to pull up,
+          // we know it's a RexInputRef, so converter isn't needed.
+          compenProjs.add(
+              new RexInputRef(i - leftCount + fieldCnt(target.getLeft()), type));
+        } else {
+          RexNode apply = converter1.apply(qInput1Projs.get(i - leftCount));
+          compenProjs.add(adjustNullability(apply, type, rexBuilder));
+        }
+      }
     }
-    return true;
+    return compenProjs;
   }
 
-  /** Returns whether all projects are nullable. */
-  private static boolean allProjectsNullable(List<RexNode> projects) {
-    return projects.stream()
-        .allMatch(project -> project.getType().isNullable());
+  /** Cast RexNode to the given type if only nullability differs. */
```

Comment:
I would also add "Otherwise throw". The current comment implies that in all other cases the type is left unchanged.